configure: add debug options asan and ubsan

hzhou opened the pull request:
## Pull Request Description

Add configure options --enable-g=asan and --enable-g=ubsan for address
sanitizer and undefined behavior sanitizer.

[skip warnings]
## Expected Impact

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ofi

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ofi

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ofi
env: MPITEST_MEMORY_TOTAL=10

hzhou commented:
test:mpich/custom --enable-g=usan
netmod: ch4:ofi

hzhou commented:
test:mpich/custom --enable-g=usan
netmod: ch4:ofi
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=usan
netmod: ch4:ofi
compiler: gcc-9

raffenet commented on src/mpi/datatype/typerep/src/typerep_yaksa_pack.c:
```diff
@@ -15,6 +15,11 @@ int MPIR_Typerep_copy(void *outbuf, const void *inbuf, MPI_Aint num_bytes)
     int mpi_errno = MPI_SUCCESS;
     int rc;
 
+    if (outbuf == NULL || inbuf == NULL) {
```
Could this branch just be for `num_bytes == 0`? That way it skips all the logic below even when valid buffers are supplied.

hzhou replied:
Yes, of course!

hzhou commented:
test:mpich/custom --enable-g=usan
netmod: ch4:ofi
compiler: gcc-9

hzhou commented:
These sanitize checks are really good that we should consider add them to review tests -- so the developer can take care of their leaks 😄

hzhou commented:
test:mpich/custom --enable-g=usan
netmod: ch4:ofi
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=usan
netmod: ch4:ofi
compiler: gcc-9

raffenet commented:
> These sanitize checks are really good that we should consider add them to review tests -- so the developers can take care of their own leaks 😄

I agree. I'm pleasantly surprised that Fortran support seems to be working this PR. I gave up on it in my Jenkins config after not too long.

hzhou commented:
> I agree. I'm pleasantly surprised that Fortran support seems to be working this PR. I gave up on it in my Jenkins config after not too long.

I think it is because here is using gcc, which works consistently with gfortran.

hzhou commented:
test:mpich/custom --enable-g=asan
compiler: clang-10

raffenet commented on src/mpid/ch4/netmod/ofi/ofi_am_impl.h:
```diff
@@ -312,6 +312,11 @@ static inline int MPIDI_OFI_am_isend_short(int rank,
     iov[2].iov_base = (void *) data;
     iov[2].iov_len = data_sz;
 
+    if (data_sz & 0x7) {
```
Couldn't this cause an out-of-bounds access at the sender? I think this issue should really be fixed in libfabric, if I understand it correctly.

hzhou replied:
Yeah, wonder why the address sanitizer didn't trigger -- maybe because it is accessed by the sockets driver, which was not instrumented. 

How about we add padding as a fourth iov?

>  I think this issue should really be fixed in libfabric, if I understand it correctly.

I agree. But libfabric won't know what alignment we need. It should have an alignment setting that we can control.

Alternatively, we could disable `FI_MULTI_RECV`.

hzhou replied:
> Yeah, wonder why the address sanitizer didn't trigger

Wait, this change is for UndefinedBehaviorSanitizer, which is made after I cleared asan tests. Maybe it will trigger now...

raffenet replied:
> I agree. But libfabric won't know what alignment we need. It should have an alignment setting that we can control.

Control would be nice to have, but defaulting to malloc-like alignment for FI_MULTI_RECV (i.e. suitably aligned for any object type) would make this a whole lot simpler for the user. It might also be good to run ubsan with libfabric+fabtests to see if the same warning can be triggered without MPICH. That would be good to have if we decide to raise the issue in their Github.

hzhou replied:
Fixing on the libfabric side is complicated. For one, it needs maintain both actual data_len and padded data_len, while not difficult but quite involved. For two, we have to fix for each provider.

On the other hand, the fix on mpich's side is rather trivial -- just pad the data. Now I used a static padding char array as an additional iov entry, the concern of accessing out-of-bound is removed.

hzhou commented:
test:mpich/custom --enable-g=asan --disable-fortran
compiler: clang-10

hzhou commented:
test:mpich/custom --enable-g=asan --disable-fortran --disable-cxx
compiler: clang-10

hzhou commented:
@raffenet Do you know what's the issue with `clang` on that cxx code: https://github.com/pmodels/mpich/pull/4657#issuecomment-649043158?

hzhou commented:
test:mpich/custom --enable-g=asan --disable-fortran --disable-cxx
netmod: ch4:ofi
compiler: clang-10

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ofi
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ofi
compiler: gcc-9
env: FI_PROVIDER=sockets

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ofi:sockets
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ofi

raffenet commented:
One thing I had 

> test:mpich/custom --enable-g=asan --disable-fortran --disable-cxx
> netmod: ch4:ofi
> compiler: clang-10
> 
> RESULT: error in cpi, no clue what is the issue:
> 
> ```
> Direct leak of 40 byte(s) in 1 object(s) allocated from:
>     #0 0x4ac9f8  (/var/lib/jenkins-slave/workspace/mpich-review-custom/examples/.libs/263177-lt-cpi (deleted)+0x4ac9f8)
>     #1 0x7fec74124df7 in vasprintf (/lib64/libc.so.6+0x77df7)
> 
> Direct leak of 7 byte(s) in 1 object(s) allocated from:
>     #0 0x435c20  (/var/lib/jenkins-slave/workspace/mpich-review-custom/examples/.libs/263177-lt-cpi (deleted)+0x435c20)
>     #1 0x7fec77bca094 in fi_ibv_init_info (/var/lib/jenkins-slave/workspace/mpich-review-custom/lib/.libs/libmpi.so.0+0x1926094)
>     #2 0x3a333062393a332f  (<unknown module>)
> 
> Direct leak of 7 byte(s) in 1 object(s) allocated from:
>     #0 0x435c20  (/var/lib/jenkins-slave/workspace/mpich-review-custom/examples/.libs/263177-lt-cpi (deleted)+0x435c20)
>     #1 0x7fec77bca094 in fi_ibv_init_info (/var/lib/jenkins-slave/workspace/mpich-review-custom/lib/.libs/libmpi.so.0+0x1926094)
>     #2 0x663a666661373a33  (<unknown module>)
> 
> SUMMARY: AddressSanitizer: 54 byte(s) leaked in 3 allocation(s).
> ```
> 
> The first leak must be leaked from one of the external libs, possibly `librdmacm.so.1` or `libibverbs.so.1`

I haven't figured this one out either. The only `vasprintf` usage I could find in `libfabric` hasn't really offered any clue.

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ucx
env: MPITEST_MEMORY_TOTAL=10

hzhou commented:
Changing the name of `usan` to `ubsan`. I believe that is more accepted reference.

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ucx
compiler: gcc-9
env: MPITEST_MEMORY_TOTAL=10

hzhou commented:
Cherry picked leak fix in libfabric and ucx--

test:mpich/custom --enable-g=asan
netmod: ch4:ucx
compiler: gcc-9
env: MPITEST_MEMORY_TOTAL=10

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ucx
compiler: gcc-9
env: MPITEST_MEMORY_TOTAL=10

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ucx
compiler: gcc-9
env: MPITEST_MEMORY_TOTAL=10

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ucx
compiler: gcc-9
env: MPITEST_MEMORY_TOTAL=10

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ofi
compiler: gcc-9
env: MPITEST_MEMORY_TOTAL=10

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ofi
compiler: gcc-9
env: MPITEST_MEMORY_TOTAL=10
env: FI_PROVIDER=sockets

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ofi
compiler: gcc-9
env: MPITEST_MEMORY_TOTAL=10
env: FI_PROVIDER=sockets

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch3:tcp
compiler: gcc-9
env: MPITEST_MEMORY_TOTAL=10
env: FI_PROVIDER=sockets

hzhou commented:
@raffenet It's ready for review.

raffenet commented on src/mpid/ch4/netmod/ofi/ofi_am_impl.h:
```diff
@@ -315,9 +316,18 @@ static inline int MPIDI_OFI_am_isend_short(int rank,
+        static char pad[16];
+        iov[3].iov_base = pad;
+        iov[3].iov_len = 8 - (total_len & 0x7);
+        num_iov = 4;
```
I thought we used to check if the provider `iov_limit` was >= 3 to ensure support for the active message APIs. However, I can't find it, so maybe it is only an implicit assumption. The bump to 4 probably doesn't make any difference.

pavanbalaji replied:
@raffenet In our discussion for GPUs, we wanted to move away from the IOV model for AMs.  It's confusing and is an artifact of what CH3 used to have.  The intent here is to send three pointers: (1) user buffer, (2) header, and (3) optional extended header.  I think those three should be explicit parameters to the function, rather than sticking them inside an IOV.  Just pointing it out in case you are considering adding even more content to the IOV.

hzhou replied:
@pavanbalaji Any idea how to address the alignment issue when the payload is not multiple of required alignment? I currently can think of two solutions:
* Don't use `FI_MULTI_RECV` -- need evaluate and determine how much benefit of `FI_MULTI_RECV` bring us.
* Stage the payload buffer -- the buffer need come form persistent allocations and the additional memcpy will hurt large payload.

hzhou replied:
Note that the suggestion for libfabric to fix the alignment issue has been turned down -- https://github.com/ofiwg/libfabric/issues/6041

pavanbalaji replied:
We could copy the header out before using it.  We did the same thing for `ch3` too.

hzhou replied:
> 
> 
> We could copy the header out before using it. We did the same thing for `ch3` too.

OK, I'll try that.

raffenet commented on src/mpid/ch4/src/ch4_impl.h:
```diff
@@ -997,8 +1005,15 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_compute_acc_op(void *source_buf, int source_
-            (*uop) ((char *) source_buf + src_type_stride * accumulated_count,
-                    (char *) target_buf + MPIR_Ptr_to_aint(curr_loc), &count, &type);
+            if (src_ptr) {
+                memcpy(src_ptr, (char *) source_buf + src_type_stride * accumulated_count,
```
Shouldn't we use `MPIR_Typerep_unpack` here? I think this will fix the issues with accumulate on pairtypes with padding.

raffenet replied:
`MPI_SHORT_INT` is currently xfailed because the accumulate ops do not account for the missing 2-byte pad in the packed version.

hzhou replied:
> 
> 
> Shouldn't we use `MPIR_Typerep_unpack` here?

Are we concerned that `source_buf` may be GPU buffer (the reason you suggested `MPIR_Typerep_unpack`)?

hzhou replied:
> Shouldn't we use `MPIR_Typerep_unpack` here? I think this will fix the issues with accumulate on pairtypes with padding.

Ahh, I was being dense. Finally I realized for pair types, it is not just omitting the padding during packing; it is literally *packed* as a non-contig data. Thus to unpack, we need use `MPIR_Typerep_unpack` -- for the sake of `MPI_SHORT_INT`.

hzhou commented:
Fixed a type, removed xfail of `atomic_get_short_int`:

test:mpich/custom --enable-g=asan
netmod: ch4:ofi
compiler: gcc-9
env: MPITEST_MEMORY_TOTAL=10
env: FI_PROVIDER=sockets

hzhou commented:
test:mpich/custom --enable-g=ubsan
netmod: ch4:ofi
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=ubsan
netmod: ch4:ofi
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=ubsan
netmod: ch4:ofi
compiler: gcc-9

hzhou commented:
Failed to reproduce. Try it again:

test:mpich/custom --enable-g=ubsan
netmod: ch4:ofi
compiler: gcc-9
testlist: f90/pt2pt/sendrecvreplf90 2

hzhou commented:
Couldn't reproduce. Try it again:

test:mpich/custom --enable-g=ubsan
netmod: ch4:ofi
compiler: gcc-9
testlist: f90/pt2pt

hzhou commented:
test:mpich/custom --enable-g=ubsan
netmod: ch4:ofi
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=ubsan
netmod: ch4:ofi
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=ubsan
netmod: ch4:ofi
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=ubsan
netmod: ch4:ofi
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=ubsan
netmod: ch4:ucx
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=asan
netmod: ch4:ucx
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=ubsan
netmod: ch4:ucx
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=ubsan
netmod: ch4:ucx
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=ubsan
netmod: ch4:ucx
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=ubsan
netmod: ch4:ucx
compiler: gcc-9

hzhou commented:
test:mpich/custom --enable-g=ubsan
netmod: ch4:ucx
compiler: gcc-9

hzhou commented:
Can't wait to merge this, running final check:

test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/custom
config: asan
netmod: ch4:ofi

raffenet commented:
I think this will be clean if we add back the timeout multiplier to the test-worker script.

hzhou commented:
> I think this will be clean if we add back the timeout multiplier to the test-worker script.

Ahh, I see. You mean the segfault can be triggered by the mpiexec timeout killer? I'll check

raffenet commented:
> > I think this will be clean if we add back the timeout multiplier to the test-worker script.
> 
> Ahh, I see. You mean the segfault can be triggered by the mpiexec timeout killer? I'll check

Oh, I just clicked on one failure and saw it was a timeout. Now I see that's not the only reason some tests fail in ch4/ofi.

hzhou commented:
> 
> Oh, I just clicked on one failure and saw it was a timeout. Now I see that's not the only reason some tests fail in ch4/ofi.

Yeah, I need check what's going on with the am path.

hzhou commented:
@raffenet Added last commit to address the ofi am (lmt) issue. Could take a look? I'll re-run the tests after merging another PR.

hzhou commented:
test:mpich/ch4/ofi

test:mpich/custom
config: asan
netmod: ch4:ofi

hzhou commented:
Added forcing `FI_PROVIDER=sockets` to jenkins script:

test:mpich/custom
config: asan
netmod: ch4:ofi

raffenet commented on src/mpid/ch4/netmod/ofi/ofi_pre.h:
```diff
@@ -101,9 +101,11 @@ typedef struct {
-    MPIDI_OFI_lmt_msg_payload_t *lmt_msg;
     void *unpack_buffer;
     MPI_Aint pack_size;
+    uint16_t src_offset;
```
This is `uint64_t` in `MPIDI_OFI_lmt_msg_payload_t`.

hzhou replied:
Ouch, thanks for checking!

raffenet replied:
👍 this version looks good.

hzhou commented:
test:mpich/custom
config: ubsan
netmod: ch4:ofi
